<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: Subclonal copy number and LOH prediction from whole genome
sequencing of tumours</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> Subclonal copy number and LOH prediction from whole genome
sequencing of tumours
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;TitanCNA&rsquo; version 1.24.0</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../doc/index.html">User guides, package vignettes and other documentation.</a></li>
<li><a href="../NEWS">Package NEWS</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="TitanCNA-package.html">TitanCNA-package</a></td>
<td>TITAN: Subclonal copy number and LOH prediction whole genome sequencing of tumours</td></tr>
<tr><td style="width: 25%;"><a href="computeSDbwIndex.html">computeSDbwIndex</a></td>
<td>Compute the S_Dbw Validity Index for 'TitanCNA' model selection</td></tr>
<tr><td style="width: 25%;"><a href="TitanCNA-dataset.html">convergeParams</a></td>
<td>TITAN EM trained results for an example dataset</td></tr>
<tr><td style="width: 25%;"><a href="correctCN.html">correctIntegerCN</a></td>
<td>Compute purity and ploidy corrected log ratios; recompute integer CN for high-level amplifications.</td></tr>
<tr><td style="width: 25%;"><a href="correctReadDepth.html">correctReadDepth</a></td>
<td>Correct GC content and mappability biases in sequencing data read counts</td></tr>
<tr><td style="width: 25%;"><a href="TitanCNA-dataset.html">data</a></td>
<td>TITAN EM trained results for an example dataset</td></tr>
<tr><td style="width: 25%;"><a href="TitanCNA-dataset.html">EMresults</a></td>
<td>TITAN EM trained results for an example dataset</td></tr>
<tr><td style="width: 25%;"><a href="filterData.html">filterData</a></td>
<td>Filter list object based on read depth and missing data and returns a filtered data.table object.</td></tr>
<tr><td style="width: 25%;"><a href="haplotype.html">getHaplotypesFromVCF</a></td>
<td>Function to load tumour allele counts from a text file or data.frame and returns a data.table ('loadHaplotypeAlleleCounts').  Function to load phased heterozygous sites from a VCF file ('getHaplotypesFromVCF')</td></tr>
<tr><td style="width: 25%;"><a href="getPositionOverlap.html">getPositionOverlap</a></td>
<td>Function to assign values to given chromosome-position that overlaps a list of chromosomal segments</td></tr>
<tr><td style="width: 25%;"><a href="loadAlleleCounts.html">loadAlleleCounts</a></td>
<td>Function to load tumour allele counts from a text file or data.frame and returns a data.table.</td></tr>
<tr><td style="width: 25%;"><a href="haplotype.html">loadBXcountsFromBEDDir</a></td>
<td>Function to load tumour allele counts from a text file or data.frame and returns a data.table ('loadHaplotypeAlleleCounts').  Function to load phased heterozygous sites from a VCF file ('getHaplotypesFromVCF')</td></tr>
<tr><td style="width: 25%;"><a href="loadDefaultParameters.html">loadDefaultParameters</a></td>
<td>Load TITAN parameters</td></tr>
<tr><td style="width: 25%;"><a href="haplotype.html">loadHaplotypeAlleleCounts</a></td>
<td>Function to load tumour allele counts from a text file or data.frame and returns a data.table ('loadHaplotypeAlleleCounts').  Function to load phased heterozygous sites from a VCF file ('getHaplotypesFromVCF')</td></tr>
<tr><td style="width: 25%;"><a href="TitanCNA-output.html">outputModelParameters</a></td>
<td>Formatting and printing 'TitanCNA' results.</td></tr>
<tr><td style="width: 25%;"><a href="TitanCNA-output.html">outputTitanResults</a></td>
<td>Formatting and printing 'TitanCNA' results.</td></tr>
<tr><td style="width: 25%;"><a href="TitanCNA-output.html">outputTitanSegments</a></td>
<td>Formatting and printing 'TitanCNA' results.</td></tr>
<tr><td style="width: 25%;"><a href="TitanCNA-plotting.html">plotAllelicRatio</a></td>
<td>Plotting functions for 'TitanCNA' results.</td></tr>
<tr><td style="width: 25%;"><a href="TitanCNA-plotting.html">plotClonalFrequency</a></td>
<td>Plotting functions for 'TitanCNA' results.</td></tr>
<tr><td style="width: 25%;"><a href="TitanCNA-plotting.html">plotCNlogRByChr</a></td>
<td>Plotting functions for 'TitanCNA' results.</td></tr>
<tr><td style="width: 25%;"><a href="TitanCNA-plotting.html">plotHaplotypeFraction</a></td>
<td>Plotting functions for 'TitanCNA' results.</td></tr>
<tr><td style="width: 25%;"><a href="TitanCNA-plotting.html">plotSegmentMedians</a></td>
<td>Plotting functions for 'TitanCNA' results.</td></tr>
<tr><td style="width: 25%;"><a href="TitanCNA-plotting.html">plotSubcloneProfiles</a></td>
<td>Plotting functions for 'TitanCNA' results.</td></tr>
<tr><td style="width: 25%;"><a href="runEMclonalCN.html">runEMclonalCN</a></td>
<td>Function to run the Expectation Maximization Algorithm in 'TitanCNA'.</td></tr>
<tr><td style="width: 25%;"><a href="loadAlleleCounts.html">setGenomeStyle</a></td>
<td>Function to load tumour allele counts from a text file or data.frame and returns a data.table.</td></tr>
<tr><td style="width: 25%;"><a href="TitanCNA-package.html">TitanCNA</a></td>
<td>TITAN: Subclonal copy number and LOH prediction whole genome sequencing of tumours</td></tr>
<tr><td style="width: 25%;"><a href="TitanCNA-dataset.html">TitanCNA-dataset</a></td>
<td>TITAN EM trained results for an example dataset</td></tr>
<tr><td style="width: 25%;"><a href="viterbiClonalCN.html">viterbiClonalCN</a></td>
<td>Function to run the Viterbi algorithm for 'TitanCNA'.</td></tr>
<tr><td style="width: 25%;"><a href="wigImport.html">wigToGRanges</a></td>
<td>WIG Import Functions. wigToGRanges (new) and wigToRangedData (deprecated)</td></tr>
<tr><td style="width: 25%;"><a href="wigImport.html">wigToRangedData</a></td>
<td>WIG Import Functions. wigToGRanges (new) and wigToRangedData (deprecated)</td></tr>
</table>
</body></html>
